Python/chapter02/2.8 - Loop Detection/miguel_soln_2.8.py

```diff
@@ -171,16 +171,20 @@ def loop_detection_linear_time_const_space(ll: LinkedList):
     A linked list is 'circular' when a node's
     next pointer points to an earlier node, so
     as to make a loop in the linked list.
+    Floyd cycle-finding algorithm:
+        https://www.geeksforgeeks.org/detect-loop-in-a-linked-list/
     Runtime:  O(n)
     Space Complexity:  O(1)
     :param ll: an input linked list
     :return: the corrupt node or None
     """
-    curr_node = ll.head
-    while curr_node is not None:
-        if curr_node is ll.tail.next:
-            return curr_node
-        curr_node = curr_node.next
+    slow_ptr = ll.head
+    fast_ptr = ll.head
+    while slow_ptr and fast_ptr and fast_ptr.next:
+        slow_ptr = slow_ptr.next
+        fast_ptr = fast_ptr.next.next
+        if slow_ptr is fast_ptr or slow_ptr is fast_ptr.next:
```

Comment:
you only need one of these comparisons. Adding both makes it harder to reason about.

Reply:
For the test case 1 -> 2 -> 3 -> 4 -> 5 -> 3
leaving out the second comparison leads to the slow and fast pointer meeting up at the node containing 4, which is not the start of the loop and caused the first test case to fail.

Reply:
i'm pretty sure you can construct a test case that makes it fail _even_ with the duplicate test. say, if the loop is long enough, it'll detect the loop somewhere in the middle.

what this means is that the cycle-finding algorithm only lets you know whether there is a cycle and _a_ node within it. you need additional logic if you want the very first.

Reply:
(and having said that, can you find such a test case?)

Reply:
Oh ok you are right,
Yes I constructed such a test case.
Going to add the logic for finding the first node in the cycle

Reply:
in general, every time you add complexity (mostly in the form of one-off tests and branches) to your algorithm, you need to analyze whether those one-off tests are needed (maybe they are covered by a more general case), and whether they are fixing a real issue or just papering over one narrow set of cases.

This is what trips 80% of the interviewees in my experience.

Reply:
updated